staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto

```diff
@@ -1512,3 +1517,68 @@ message ReopenContainerLogRequest {
 
 message ReopenContainerLogResponse{
 }
+
+// Common options used for checkpointing and restoring.
+message CheckpointRestoreOptions {
+    // Keep temporary files. Like log files. Helpful for debugging.
+    bool keep_temporary_files = 1;
+    // Checkpoint/Restore the container with established TCP connections.
+    bool keep_network_state = 2;
+    // Location of the checkpoint archive used for export/import
+    string archive_location = 3;
+    // Instead of using a local checkpoint archive use a OCI image
+    // to store the checkpoint.
+    bool archive_is_oci_image = 4;
+    // Specify the checkpoint archive compression algorithm
+    // See github.com/containers/storage/pkg/archive/archive.go for
+    // possible values.
+    int64 compression = 5;
+}
+
+message CheckpointContainerOptions {
+    CheckpointRestoreOptions common_options = 1;
+    // Leave the container running after checkpointing
+    bool leave_running = 2;
+}
+
+message CheckpointContainerRequest {
+    // ID of the container to be checkpointed.
+    string container_id = 1;
+    // Options for checkpointing.
+    CheckpointContainerOptions options = 2;
+}
+
+message CheckpointContainerResponse {
+}
+
+message RestoreContainerOptions {
+    CheckpointRestoreOptions common_options = 1;
+    // ID of the sandbox into which the container should be restored
+    // If this is not specified the container will be restored
+    // in the sandbox it was checkpointed from.
+    string pod_sandbox_id = 2;
+    // Restoring multiple copies of a checkpointed container may
+    // require a new name for the restored container.
+    string name = 3;
+    // The labels and annotations are in the same form as in ContainerConfig
+    // during CreateContainerRequest.
+    map<string, string> labels = 4;
+    map<string, string> annotations = 5;
+    // Restoring a container in another pod might require changing
+    // Kubernetes specific mounts if the mount source depends on
+    // the pod ID.
+    // This tells the container engine to replace existing mount
+    // sources with a new source.
+    map<string, string> change_mounts = 6;
```

Comment:
Can you give an example of how this may be used?

Reply:
That is needed to restore a container outside of Kubernetes. Especially for the forensic use case it is important.

Each container has a couple of Kubernetes specific bind mounts in the container. One example is `/etc/hosts`. In one of my examples it is bind mounted from `/var/lib/kubelet/pods/3f24e9ed-78e5-48ad-b300-0a10ba5bb181/etc-hosts`. During restore, on another node, potentially without Kubernetes CRIU will try to restore the bind mount from `/var/lib/kubelet/pods/3f24e9ed-78e5-48ad-b300-0a10ba5bb181/etc-hosts` to `/etc/hosts` and as the container external file is missing, the restore will fail.

With this option in the CRI I can tell the container engine via crictl to bind mount another file as `/etc/hosts` from the outside of the container.

I kept this one in, as it essential for the restore part outside of Kubernetes for the forensic use case.

All other currently unneeded CRI entries have been removed.

Reply:
@adrianreber Couldn't we add some hard-coded list here for well-known mount path patterns so we can do this ID replacement in the restore code in the runtime?

Reply:
hard-coded does not sound doable, but I removed it and I will provide a tool to adapt the checkpoint archive. That way we can change the mount points after checkpointing but before restoring. It will result in a manual extra step before being able to restore it. That way we can learn if it would be helpful/important to have it in the CRI API or if it works with an additional step.